### PR TITLE
fix(RemoteStore): avoid listing all objects in remote store in empty() method

### PR DESCRIPTION
**Pull request by jhamman**
Some initial exploration with the RemoteStore found that the `empty()` check was taking a very long time. This sped it up considerably. 

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: src/zarr/storage/remote.py [hunk 1]
```diff
@@ -94,7 +94,11 @@ async def clear(self) -> None:
             pass
 
     async def empty(self) -> bool:
-        return not await self.fs._find(self.path, withdirs=True)
+        async for _path, _dirs, files in self.fs._walk(self.path):
+            # stop once a file is found
+            if files:
```
Comment (martindurant) [inline src/zarr/storage/remote.py:99]: Do included directories count as content? The original implied yes.

File: src/zarr/storage/remote.py [hunk 2]
```diff
@@ -94,7 +94,11 @@ async def clear(self) -> None:
             pass
 
     async def empty(self) -> bool:
-        return not await self.fs._find(self.path, withdirs=True)
+        async for _path, _dirs, files in self.fs._walk(self.path):
```
Comment (martindurant) [inline src/zarr/storage/remote.py:97]: It cannot be guaranteed that this doesn't do a find() too... Would `_ls()` be enough?

There was talk of making `ls()` and friends in fsspec be generators, but not yet.

File: src/zarr/storage/remote.py (hunk 1 above)
Comment (jhamman) [reply src/zarr/storage/remote.py:99]: We only check for files in the `LocalStore`:

https://github.com/zarr-developers/zarr-python/blob/5a134bf2c1c5e6655edd429ca554aa9699327463/src/zarr/storage/local.py#L101-L111

File: src/zarr/storage/remote.py (hunk 2 above)
Comment (jhamman) [reply src/zarr/storage/remote.py:97]: gotcha... This was an improvement using gcsfs but I can see how it wouldn't be for others. I'll take your lead on the best path forward here. All we need to know is if there are any files in the store (ignoring empty directories).

Comment (martindurant) [reply src/zarr/storage/remote.py:97]: Given the very many files that can in theory be in a zarr hierarchy, maybe we can keep with this strategy for now. Having said that, in the zarr case, we do know which subdir ought to have the fewest items in it, right?

**Review by d-v-b — APPROVED** [review]
thanks for this. perennial question: what kind of performance monitoring should we have in place to flag a regression here (or anywhere)?

**Pull request merged**